lib/maths/CTimeSeriesTestForSeasonality.cc

```diff
@@ -112,357 +175,707 @@ std::string CNewSeasonalComponentSummary::print() const {
-                              CBasicStatistics::mean(valuesMinusTrend[i])};
-                CBasicStatistics::moment<0>(decomposition.s_TrendInitialValues[i]) = offset;
-            }
-            decompositions.push_back(std::move(decomposition));
+
+    // Weekday/weekend modulation removing trend after determining decomposition.
+    m_TemporaryValues = m_Values;
+    m_CandidatePeriods = CSignal::tradingDayDecomposition(
+        m_TemporaryValues, m_OutlierFraction, this->week(), m_StartOfWeekOverride);
+    m_CandidatePeriods.push_back(CSignal::seasonalComponentSummary(this->year()));
+    this->removeIfNotTestable(m_CandidatePeriods);
+    if (m_CandidatePeriods.size() > 0 && // Did we find candidate weekend/weekday split?
+        this->includesPermittedPeriod(m_CandidatePeriods) &&
+        this->alreadyModelled(m_CandidatePeriods) == false &&
+        removeTrend(m_CandidatePeriods, m_ValuesMinusTrend, m_ModelTrendSegments)) {
+        this->testAndAddDecomposition(m_CandidatePeriods, m_ModelTrendSegments,
+                                      m_ValuesMinusTrend, false, decompositions);
+    }
+
+    // Weekday/weekend modulation removing trend before determining decomposition.
+    if (removeTrend({}, m_ValuesMinusTrend, m_ModelTrendSegments)) {
+        m_CandidatePeriods = CSignal::tradingDayDecomposition(
+            m_TemporaryValues, m_OutlierFraction, this->week(), m_StartOfWeekOverride);
+        m_CandidatePeriods.push_back(CSignal::seasonalComponentSummary(this->year()));
+        this->removeIfNotTestable(m_CandidatePeriods);
+        if (m_CandidatePeriods.size() > 0 && // Did we find candidate weekend/weekday split?
+            m_CandidatePeriods != m_Periods && // Have we already tested these periods?
+            this->includesPermittedPeriod(m_CandidatePeriods) &&
+            this->alreadyModelled(m_CandidatePeriods) == false) {
+            this->testAndAddDecomposition(m_CandidatePeriods, m_ModelTrendSegments,
+                                          m_ValuesMinusTrend, false, decompositions);
         }
     }
+
+    // Day, week, year.
+    m_Periods = m_CandidatePeriods;
+    m_CandidatePeriods.assign({CSignal::seasonalComponentSummary(this->day()),
+                               CSignal::seasonalComponentSummary(this->week()),
+                               CSignal::seasonalComponentSummary(this->year())});
+    this->removeIfNotTestable(m_CandidatePeriods);
+    if (m_CandidatePeriods.size() > 0 && // Is there sufficient data?
+        m_CandidatePeriods != m_Periods && // Have we already tested these periods?
+        this->includesPermittedPeriod(m_CandidatePeriods) &&
+        this->alreadyModelled(m_CandidatePeriods) == false &&
+        removeTrend(m_CandidatePeriods, m_ValuesMinusTrend, m_ModelTrendSegments)) {
+        this->testAndAddDecomposition(m_CandidatePeriods, m_ModelTrendSegments,
+                                      m_ValuesMinusTrend, false, decompositions);
+    }
```

Comment:
seems like this could be a subroutine.

Reply:
Done in [827cd2b](https://github.com/elastic/ml-cpp/pull/1451/commits/827cd2ba61dde5243ca42b851876846a2e55ba8f).